shell/platform/darwin/macos/framework/Source/FlutterEmbedderKeyResponder.mm

```diff
@@ -0,0 +1,715 @@
+ *
+ * The flags compared are all flags after masking with
+ * |modifierFlagOfInterestMask| and excluding |ignoringFlags|.
+ */
+- (void)synchronizeModifiers:(NSUInteger)currentFlags
+               ignoringFlags:(NSUInteger)ignoringFlags
+                   timestamp:(NSTimeInterval)timestamp;
+
+/**
+ * Update the pressing state.
+ *
+ * If `logicalKey` is not 0, `physicalKey` is pressed as `logicalKey`.
+ * Otherwise, `physicalKey` is released.
+ */
+- (void)updateKey:(uint64_t)physicalKey asPressed:(uint64_t)logicalKey;
+
+/**
+ * Send an event to the framework, expecting its response.
+ */
+- (void)sendPrimaryFlutterEvent:(const FlutterKeyEvent&)event
+                       callback:(nonnull FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Send a CapsLock down event, then a CapsLock up event.
+ *
+ * If downCallback is nil, then both events will be synthesized. Otherwise, the
+ * downCallback will be used as the callback for the down event, which is not
+ * synthesized.
+ */
+- (void)sendCapsLockTapWithTimestamp:(NSTimeInterval)timestamp
+                            callback:(nullable FlutterKeyCallbackGuard*)downCallback;
+
+/**
+ * Send a key event for a modifier key.
+ *
+ * If callback is nil, then the event is synthesized.
+ */
+- (void)sendModifierEventOfType:(BOOL)isDownEvent
+                      timestamp:(NSTimeInterval)timestamp
+                        keyCode:(unsigned short)keyCode
+                       callback:(nullable FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Processes a down event.
+ */
+- (void)handleDownEvent:(nonnull NSEvent*)event callback:(nonnull FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Processes an up event.
+ */
+- (void)handleUpEvent:(nonnull NSEvent*)event callback:(nonnull FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Processes an up event.
+ */
+- (void)handleCapsLockEvent:(nonnull NSEvent*)event
+                   callback:(nonnull FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Processes a flags changed event, where modifier keys are pressed or released.
+ */
+- (void)handleFlagEvent:(nonnull NSEvent*)event callback:(nonnull FlutterKeyCallbackGuard*)callback;
+
+/**
+ * Processes the response from the framework.
+ */
+- (void)handleResponse:(BOOL)handled forId:(uint64_t)responseId;
+
+@end
+
+@implementation FlutterEmbedderKeyResponder
+
+- (nonnull instancetype)initWithSendEvent:(FlutterSendEmbedderKeyEvent)sendEvent {
+  self = [super init];
+  if (self != nil) {
+    _sendEvent = sendEvent;
+    _pressingRecords = [NSMutableDictionary dictionary];
+    _pendingResponses = [NSMutableDictionary dictionary];
+    _responseId = 1;
+    _lastModifierFlags = 0;
+    _modifierFlagOfInterestMask = computeModifierFlagOfInterestMask();
+  }
+  return self;
+}
+
+- (void)handleEvent:(NSEvent*)event callback:(FlutterAsyncKeyCallback)callback {
+  // The conversion algorithm relies on a non-nil callback to properly compute
+  // `synthesized`. If someday callback is allowed to be nil, make a dummy empty
```

Comment:
I'm not sure you need the "If someday" sentence at all: it's purely speculative, and it'll be obvious at that time what to do.